csrc/multidevice/collective.cpp

```diff
@@ -0,0 +1,199 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+
+#include <multidevice/collective.h>
+
+namespace nvfuser {
+namespace {
+static inline std::vector<std::vector<at::Tensor>> setBufList(
+    std::vector<at::Tensor> bufs) {
+  return {std::move(bufs)};
+}
+
+static inline void assertBufferCount(
+    const std::vector<at::Tensor>& bufs,
+    size_t count) {
+  TORCH_INTERNAL_ASSERT(
+      bufs.size() == count,
+      "there must be ",
+      count,
+      " buffer(s), but ",
+      bufs.size(),
+      " were given");
+}
+
+static inline void assertBuffersHaveSameSize(
+    const std::vector<at::Tensor>& bufs1,
+    const std::vector<at::Tensor>& bufs2) {
+  if (bufs1.empty() && bufs2.empty()) {
+    return;
+  }
+  auto sizes = (bufs1.empty() ? bufs2 : bufs1).at(0).sizes();
+  for (auto& bufs : {bufs1, bufs2}) {
+    for (auto& buf : bufs) {
+      TORCH_INTERNAL_ASSERT(
+          buf.sizes() == sizes, "all buffers must have the same size");
+    }
+  }
+}
+
+static inline void post_common(Collective& self, Communicator& comm) {
+  TORCH_INTERNAL_ASSERT(
+      std::find(
+          self.params().team.begin(),
+          self.params().team.end(),
+          comm.deviceId()) != self.params().team.end(),
+      "current device index ",
+      comm.deviceId(),
+      " must be present in the collective's team");
+}
+
+} // namespace
+
+Collective::Collective(CollectiveParams params, std::string name, bool has_root)
+    : params_(std::move(params)),
+      collective_type_(std::move(name)),
+      has_root_(has_root) {
+  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
```

Comment:
Why does the source buffer need to have the same shape as the destination buffer?

Reply:
I am not sure to understand. The buffer total lengths need to match by definition.
This is later checked by nccl [here ](https://github.com/pytorch/pytorch/blob/e1aba2c8c39d817da391a75aa7ce6dd590f93755/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1527C14-L1527C14)

Reply:
Maybe this is related to https://github.com/NVIDIA/Fuser/pull/797/files#r1316555803.

Let's say gathering a single byte from all devices. Each device should have just a one-byte source buffer. I thought only the root device should have a single flattened destination buffer, so its size should be equal to the number of devices.

It seems buffers must be unflattend, so this assertion seems to make sense.

Reply:
> Let's say gathering a single byte from all devices. Each device should have just a one-byte source buffer. I thought only the root device should have a single flattened destination buffer, so its size should be equal to the number of devices.

Every device will have a one-byte src buffer. The root will have <nbr_of_devices> * one-byte dst buffers (and not one big buffer), the others will have no dst buffer.
Is it clearer ?

Reply:
Yeah, this was different from what I expected to see with communication APIs. I don't think it's a good design, but 🤷